webrtc/stun/stun_connection.nim

```diff
@@ -0,0 +1,249 @@
+        await self.dataRecv.addLast(message)
+        continue
+      if decoded.msgType == StunBindingErrorResponse:
+        trace "Received a STUN error", decoded, remote = self.raddr
+        continue
+      elif decoded.msgType == StunBindingResponse:
+        # TODO: Process StunBindingResponse doesn't seem necessary for libp2p-webrtc-direct.
+        # Some browsers could be uncooperative. In that case, it should be implemented.
+        # It should be implemented for libp2p-webrtc.
+        continue
+      elif decoded.msgType == StunBindingRequest:
+        let errorOpt = self.checkForError(decoded)
+        if errorOpt.isSome():
+          let error = errorOpt.get()
+          await self.conn.write(self.raddr, error.encode(@[]))
+          continue
+
+        let
+          bindingResponse = self.getBindingResponse(decoded)
+          usernameAttr = decoded.getAttribute(UsernameAttribute).get()
+          password = self.passwordProvider(usernameAttr.username)
+        await self.conn.write(
+          self.raddr,
+          bindingResponse.encode(password)
+        )
+    except SerializationError as exc:
+      warn "Failed to decode the Stun message", error=exc.msg, message
+    except WebRtcError as exc:
+      trace "Failed to write the Stun response", error=exc.msg
+
+proc init*(
+    T: type StunConn,
+    conn: UdpConn,
+    raddr: TransportAddress,
+    iceControlling: bool,
+    usernameProvider: StunUsernameProvider,
+    usernameChecker: StunUsernameChecker,
+    passwordProvider: StunPasswordProvider,
+    rng: ref HmacDrbgContext
+  ): T =
+  ## Initialize a Stun Connection
+  ## `conn` the underlying Udp Connection
+  ## `raddr` the remote address observed while receiving message with Udp
+  ## `iceControlling` flag to know if we're supposed to act as a "client"
+  ##   (controlling) or a "server" (controlled)
+  ## `usernameProvider` callback to get a username for the Username attribute
+  ## `usernameChecker` callback to let the user check if the Username received
+  ##   is valid or not
+  ## `passwordProvider` callback to get a key password for the
+  ##   Message-integrity sha1 encryption
+  ##
+  var self = T(
+    conn: conn,
+    laddr: conn.laddr,
+    raddr: raddr,
+    closed: false,
+    closeEvent: newAsyncEvent(),
+    dataRecv: newAsyncQueue[seq[byte]](),
+    stunMsgs: newAsyncQueue[seq[byte]](),
+    iceControlling: iceControlling,
+    iceTiebreaker: rng[].generate(uint32),
+    usernameProvider: usernameProvider,
+    usernameChecker: usernameChecker,
+    passwordProvider: passwordProvider,
+    rng: rng
+  )
+  self.handlesFut = self.stunMessageHandler()
+  return self
+
+proc join*(self: StunConn) {.async: (raises: [CancelledError]).} =
+  ## Wait for the Stun Connection to be closed
+  ##
+  await self.closeEvent.wait()
+
+proc close*(self: StunConn) =
+  ## Close a Stun Connection
+  ##
+  if self.closed:
+    debug "Try to close an already closed StunConn"
+    return
+  self.closed = true
```

Comment:
should it be the last line?

Reply:
Or maybe having a `closing` and `closed`, but not sure.

Reply:
I changed it, but as the proc is synchronous, it doesn't change anything